File: src/Autototp/Services/ImportService.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Autototp.Models;

namespace Autototp.Services;

public sealed class ImportedAccount : System.ComponentModel.INotifyPropertyChanged
{
    private bool _isSelected = true;
    private string _windowTitleMatch = string.Empty;

    public string Name { get; init; } = string.Empty;
    public string Issuer { get; init; } = string.Empty;
    public string Secret { get; init; } = string.Empty;
    public int Digits { get; init; } = 6;
    public int Period { get; init; } = 30;
    public string Algorithm { get; init; } = "SHA1";

    public string WindowTitleMatch
    {
        get => _windowTitleMatch;
        set
        {
            if (_windowTitleMatch == value)
            {
                return;
            }

            _windowTitleMatch = value;
            PropertyChanged?.Invoke(this, new System.ComponentModel.PropertyChangedEventArgs(nameof(WindowTitleMatch)));
        }
    }

    public bool IsSelected
    {
        get => _isSelected;
        set
        {
            if (_isSelected == value)
            {
                return;
            }

            _isSelected = value;
            PropertyChanged?.Invoke(this, new System.ComponentModel.PropertyChangedEventArgs(nameof(IsSelected)));
        }
    }

    public event System.ComponentModel.PropertyChangedEventHandler? PropertyChanged;
}

public sealed class ImportResult
{
    public List<ImportedAccount> Accounts { get; init; } = [];
    public List<string> Warnings { get; init; } = [];
}

public sealed class ImportService
{
    public ImportResult ParseOtpAuthText(string text)
    {
        var result = new ImportResult();
        if (string.IsNullOrWhiteSpace(text))
        {
            result.Warnings.Add("Keine otpauth:// URIs gefunden.");
            return result;
        }

        foreach (var rawLine in text.Split(['\r', '\n'], StringSplitOptions.RemoveEmptyEntries))
        {
            var line = rawLine.Trim();
            if (line.Length == 0 || line.StartsWith('#'))
            {
                continue;
            }

            if (!line.StartsWith("otpauth://", StringComparison.OrdinalIgnoreCase))
            {
                result.Warnings.Add($"Übersprungen (kein otpauth://): {TrimForDisplay(line)}");
                continue;
            }

            if (TryParseOtpAuthUri(line, out var account, out var warning))
            {
                result.Accounts.Add(account);
            }
            else if (!string.IsNullOrEmpty(warning))
            {
                result.Warnings.Add(warning);
            }
        }

        if (result.Accounts.Count == 0 && result.Warnings.Count == 0)
        {
            result.Warnings.Add("Keine importierbaren TOTP-Einträge gefunden.");
        }

        return result;
    }

    public ImportResult Parse2FasFile(string path, string? password = null)
    {
        var json = File.ReadAllText(path);
        return Parse2FasJson(json, password);
    }

    public ImportResult Parse2FasJson(string json, string? password = null)
    {
        var result = new ImportResult();
        using var document = JsonDocument.Parse(json);
        var root = document.RootElement;

        JsonElement services = default;
        var hasServices = root.TryGetProperty("services", out services)
            && services.ValueKind == JsonValueKind.Array
            && services.GetArrayLength() > 0;

        if (!hasServices
            && root.TryGetProperty("servicesEncrypted", out var encrypted)
            && encrypted.ValueKind == JsonValueKind.String
            && !string.IsNullOrWhiteSpace(encrypted.GetString()))
        {
            if (string.IsNullOrWhiteSpace(password))
            {
                throw new InvalidOperationException(
                    "Diese 2FAS-Sicherung ist passwortgeschützt. Bitte das Backup-Passwort eingeben.");
            }

            var decrypted = Decrypt2FasServices(encrypted.GetString()!, password);
            using var inner = JsonDocument.Parse(decrypted);
            if (inner.RootElement.ValueKind != JsonValueKind.Array)
            {
                throw new InvalidOperationException("Das entschlüsselte 2FAS-Backup hat ein unerwartetes Format.");
            }

            services = inner.RootElement.Clone();
            hasServices = true;
        }

        if (!hasServices)
        {
            result.Warnings.Add("Die 2FAS-Datei enthält keine Services.");
            return result;
        }

        foreach (var service in services.EnumerateArray())
        {
            if (TryMap2FasService(service, out var account, out var warning))
            {
                result.Accounts.Add(account);
            }
            else if (!string.IsNullOrEmpty(warning))
            {
                result.Warnings.Add(warning);
            }
        }

        return result;
    }

    public bool TryParseOtpAuthUri(string uri, out ImportedAccount account, out string warning)
    {
        account = new ImportedAccount();
        warning = string.Empty;

        if (!Uri.TryCreate(uri, UriKind.Absolute, out var parsed)
            || !parsed.Scheme.Equals("otpauth", StringComparison.OrdinalIgnoreCase))
        {
            warning = "Ungültige otpauth:// URI.";
            return false;
        }

        var type = parsed.Host;
        if (!type.Equals("totp", StringComparison.OrdinalIgnoreCase))
        {
            warning = $"Übersprungen ({type}, nur TOTP wird unterstützt): {TrimForDisplay(uri)}";
            return false;
        }

        var query = ParseQuery(parsed.Query);
        if (!query.TryGetValue("secret", out var secret) || string.IsNullOrWhiteSpace(secret))
        {
            warning = "otpauth:// URI ohne Secret.";
            return false;
        }

        secret = TotpGenerator.NormalizeSecret(secret);
        if (!TotpGenerator.TryValidateSecret(secret, out _))
        {
            warning = $"Ungültiges Secret in URI: {TrimForDisplay(uri)}";
            return false;
        }

        var label = Uri.UnescapeDataString(parsed.AbsolutePath.Trim('/'));
        var issuerFromQuery = query.GetValueOrDefault("issuer");
        SplitLabel(label, out var labelIssuer, out var labelAccount);

        var issuer = FirstNonEmpty(issuerFromQuery, labelIssuer);
        var name = FirstNonEmpty(labelAccount, label, issuer, "Account");

        account = new ImportedAccount
        {
            Name = name,
            Issuer = issuer,
            Secret = secret,
            WindowTitleMatch = issuer,
            Digits = ParseInt(query.GetValueOrDefault("digits"), 6),
            Period = ParseInt(query.GetValueOrDefault("period"), 30),
            Algorithm = (query.GetValueOrDefault("algorithm") ?? "SHA1").ToUpperInvariant(),
        };
        return true;
    }

    private static bool TryMap2FasService(JsonElement service, out ImportedAccount account, out string warning)
    {
        account = new ImportedAccount();
        warning = string.Empty;

        var otp = service.TryGetProperty("otp", out var otpEl) ? otpEl : default;
        var tokenType = ReadString(otp, "tokenType");
        if (!string.IsNullOrEmpty(tokenType) && !tokenType.Equals("TOTP", StringComparison.OrdinalIgnoreCase))
        {
            warning = $"Übersprungen ({tokenType}): {ReadString(service, "name")}";
            return false;
        }

        var secret = TotpGenerator.NormalizeSecret(ReadString(service, "secret"));
        if (!TotpGenerator.TryValidateSecret(secret, out _))
        {
            warning = $"Ungültiges Secret: {ReadString(service, "name")}";
            return false;
        }

        var name = FirstNonEmpty(
            ReadString(service, "name"),
            ReadString(otp, "account"),
            ReadString(otp, "issuer"),
            "Account");
        var issuer = FirstNonEmpty(ReadString(otp, "issuer"), name);

        account = new ImportedAccount
        {
            Name = name,
            Issuer = issuer,
            Secret = secret,
            WindowTitleMatch = issuer,
            Digits = ReadInt(otp, "digits", 6),
            Period = ReadInt(otp, "period", 30),
            Algorithm = FirstNonEmpty(ReadString(otp, "algorithm"), "SHA1").ToUpperInvariant(),
        };
        return true;
    }

    private static string Decrypt2FasServices(string servicesEncrypted, string password)
    {
        var parts = servicesEncrypted.Split(':');
        if (parts.Length < 3)
        {
            throw new InvalidOperationException("Das verschlüsselte 2FAS-Format wird nicht erkannt.");
        }

        var cipher = Convert.FromBase64String(parts[0]);
        var salt = Convert.FromBase64String(parts[1]);
        var iv = Convert.FromBase64String(parts[2]);

        using var kdf = new Rfc2898DeriveBytes(password, salt, 10_000, HashAlgorithmName.SHA256);
        var key = kdf.GetBytes(32);

        using var aes = Aes.Create();
        aes.Key = key;
        aes.IV = iv;
        aes.Mode = CipherMode.CBC;
        aes.Padding = PaddingMode.PKCS7;

        using var decryptor = aes.CreateDecryptor();
        var plain = decryptor.TransformFinalBlock(cipher, 0, cipher.Length);
        return Encoding.UTF8.GetString(plain);
    }

    private static Dictionary<string, string> ParseQuery(string query)
    {
        var map = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (string.IsNullOrEmpty(query))
        {
            return map;
        }

        var trimmed = query.StartsWith('?') ? query[1..] : query;
        foreach (var pair in trimmed.Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var idx = pair.IndexOf('=');
            if (idx < 0)
            {
                continue;
            }

            var key = Uri.UnescapeDataString(pair[..idx]);
            var value = Uri.UnescapeDataString(pair[(idx + 1)..]);
            map[key] = value;
        }

        return map;
    }

    private static void SplitLabel(string label, out string issuer, out string account)
    {
        issuer = string.Empty;
        account = label;
        var idx = label.IndexOf(':');
        if (idx > 0)
        {
            issuer = label[..idx];
            account = label[(idx + 1)..];
        }
    }

    private static string ReadString(JsonElement element, string name)
    {
        if (element.ValueKind is JsonValueKind.Undefined or JsonValueKind.Null)
        {
            return string.Empty;
        }

        return element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString() ?? string.Empty
            : string.Empty;
    }

    private static int ReadInt(JsonElement element, string name, int fallback)
    {
        if (element.ValueKind is JsonValueKind.Undefined or JsonValueKind.Null)
        {
            return fallback;
        }

        if (!element.TryGetProperty(name, out var value))
        {
            return fallback;
        }

        return value.ValueKind switch
        {
            JsonValueKind.Number when value.TryGetInt32(out var n) => n,
            JsonValueKind.String when int.TryParse(value.GetString(), out var n) => n,
            _ => fallback,
        };
    }

    private static int ParseInt(string? text, int fallback) =>
        int.TryParse(text, out var n) ? n : fallback;

    private static string FirstNonEmpty(params string?[] values) =>
        values.FirstOrDefault(v => !string.IsNullOrWhiteSpace(v))?.Trim() ?? string.Empty;

    private static string TrimForDisplay(string text) =>
        text.Length <= 48 ? text : text[..45] + "...";
}

File: macos/Sources/AutototpCore/ImportService.swift
import CommonCrypto
import CryptoKit
import Foundation

public struct ImportedAccount: Identifiable, Equatable {
    public let id = UUID()
    public var name = ""
    public var accountLogin = ""
    public var issuer = ""
    public var secret = ""
    public var windowTitleMatch = ""
    public var digits = 6
    public var period = 30
    public var algorithm = "SHA1"
    public var isSelected = true
    /// Set by the caller when the secret already exists in the store.
    public var isDuplicate = false

    public init() {}
}

public struct ImportResult {
    public var accounts: [ImportedAccount] = []
    public var warnings: [String] = []

    public init() {}
}

public enum ImportError: LocalizedError, Equatable {
    case passwordRequired
    case wrongPassword
    case unknownEncryptedFormat
    case invalidJSON

    public var errorDescription: String? {
        switch self {
        case .passwordRequired:
            return "Diese 2FAS-Sicherung ist passwortgeschützt. Bitte das Backup-Passwort eingeben."
        case .wrongPassword:
            return "Das 2FAS-Backup konnte nicht entschlüsselt werden. Ist das Passwort richtig?"
        case .unknownEncryptedFormat:
            return "Das verschlüsselte 2FAS-Format wird nicht erkannt."
        case .invalidJSON:
            return "Die Datei ist kein gültiges 2FAS-Backup."
        }
    }
}

public enum ImportService {
    // MARK: otpauth://

    public static func parseOtpAuthText(_ text: String) -> ImportResult {
        var result = ImportResult()
        for rawLine in text.components(separatedBy: .newlines) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            if line.isEmpty || line.hasPrefix("#") {
                continue
            }
            guard line.lowercased().hasPrefix("otpauth://") else {
                result.warnings.append("Übersprungen (kein otpauth://): \(trimForDisplay(line))")
                continue
            }
            switch parseOtpAuthURI(line) {
            case .success(let account):
                result.accounts.append(account)
            case .failure(let warning):
                result.warnings.append(warning.message)
            }
        }

        if result.accounts.isEmpty && result.warnings.isEmpty {
            result.warnings.append("Keine otpauth:// URIs gefunden.")
        }
        return result
    }

    public struct Warning: Error {
        public let message: String
    }

    public static func parseOtpAuthURI(_ uri: String) -> Result<ImportedAccount, Warning> {
        let escaped = uri.replacingOccurrences(of: " ", with: "%20")
        guard let components = URLComponents(string: escaped),
              components.scheme?.lowercased() == "otpauth" else {
            return .failure(Warning(message: "Ungültige otpauth:// URI."))
        }

        let type = components.host ?? ""
        guard type.lowercased() == "totp" else {
            return .failure(Warning(message: "Übersprungen (\(type), nur TOTP wird unterstützt): \(trimForDisplay(uri))"))
        }

        var query: [String: String] = [:]
        for item in components.queryItems ?? [] {
            query[item.name.lowercased()] = item.value ?? ""
        }

        guard let rawSecret = query["secret"], !rawSecret.trimmingCharacters(in: .whitespaces).isEmpty else {
            return .failure(Warning(message: "otpauth:// URI ohne Secret."))
        }
        let secret = TOTP.normalizeSecret(rawSecret)
        guard TOTP.validationError(secret) == nil else {
            return .failure(Warning(message: "Ungültiges Secret in URI: \(trimForDisplay(uri))"))
        }

        let label = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        var labelIssuer = ""
        var labelAccount = label
        if let colon = label.firstIndex(of: ":"), colon != label.startIndex {
            labelIssuer = String(label[..<colon])
            labelAccount = String(label[label.index(after: colon)...])
        }

        let issuer = firstNonEmpty(query["issuer"], labelIssuer)
        var account = ImportedAccount()
        account.name = firstNonEmpty(labelAccount, label, issuer, "Account")
        account.accountLogin = labelAccount.trimmingCharacters(in: .whitespaces)
        account.issuer = issuer
        account.secret = secret
        account.windowTitleMatch = issuer
        account.digits = Int(query["digits"] ?? "") ?? 6
        account.period = Int(query["period"] ?? "") ?? 30
        account.algorithm = (query["algorithm"] ?? "SHA1").uppercased()
        return .success(account)
    }

    // MARK: 2FAS

    public static func parse2FasJSON(_ data: Data, password: String?) throws -> ImportResult {
        guard let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            throw ImportError.invalidJSON
        }

        var services = root["services"] as? [[String: Any]] ?? []
        if services.isEmpty,
           let encrypted = root["servicesEncrypted"] as? String,
           !encrypted.trimmingCharacters(in: .whitespaces).isEmpty {
            guard let password, !password.isEmpty else {
                throw ImportError.passwordRequired
            }
            let plain = try decrypt2FasServices(encrypted, password: password)
            guard let decoded = (try? JSONSerialization.jsonObject(with: plain)) as? [[String: Any]] else {
                throw ImportError.wrongPassword
            }
            services = decoded
        }

        var result = ImportResult()
        guard !services.isEmpty else {
            result.warnings.append("Die 2FAS-Datei enthält keine Services.")
            return result
        }

        for service in services {
            switch map2FasService(service) {
            case .success(let account):
                result.accounts.append(account)
            case .failure(let warning):
                result.warnings.append(warning.message)
            }
        }
        return result
    }

    /// Whether a 2FAS file needs a password before it can be parsed.
    public static func is2FasEncrypted(_ data: Data) -> Bool {
        guard let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return false
        }
        let services = root["services"] as? [Any] ?? []
        return services.isEmpty && (root["servicesEncrypted"] as? String)?.isEmpty == false
    }

    static func map2FasService(_ service: [String: Any]) -> Result<ImportedAccount, Warning> {
        let otp = service["otp"] as? [String: Any] ?? [:]
        let serviceName = string(service, "name")
        let tokenType = string(otp, "tokenType")
        if !tokenType.isEmpty && tokenType.uppercased() != "TOTP" {
            return .failure(Warning(message: "Übersprungen (\(tokenType)): \(serviceName)"))
        }

        let secret = TOTP.normalizeSecret(string(service, "secret"))
        guard TOTP.validationError(secret) == nil else {
            return .failure(Warning(message: "Ungültiges Secret: \(serviceName)"))
        }

        let name = firstNonEmpty(serviceName, string(otp, "account"), string(otp, "issuer"), "Account")
        var account = ImportedAccount()
        account.name = name
        account.accountLogin = firstNonEmpty(string(otp, "account"), string(service, "account"))
        account.issuer = firstNonEmpty(string(otp, "issuer"), name)
        account.secret = secret
        account.windowTitleMatch = account.issuer
        account.digits = int(otp, "digits") ?? 6
        account.period = int(otp, "period") ?? 30
        account.algorithm = firstNonEmpty(string(otp, "algorithm"), "SHA1").uppercased()
        return .success(account)
    }

    /// 2FAS stores `base64(cipher+tag):base64(salt):base64(iv)`, PBKDF2-SHA256 (10k) + AES-256-GCM.
    /// AES-CBC is tried as a fallback for older exports.
    static func decrypt2FasServices(_ encrypted: String, password: String) throws -> Data {
        let parts = encrypted.split(separator: ":").map(String.init)
        guard parts.count >= 3,
              let cipher = Data(base64Encoded: parts[0]),
              let salt = Data(base64Encoded: parts[1]),
              let iv = Data(base64Encoded: parts[2]) else {
            throw ImportError.unknownEncryptedFormat
        }

        let key = pbkdf2SHA256(password: password, salt: salt, iterations: 10_000, length: 32)

        if cipher.count > 16,
           let nonce = try? AES.GCM.Nonce(data: iv),
           let box = try? AES.GCM.SealedBox(nonce: nonce, ciphertext: cipher.dropLast(16), tag: cipher.suffix(16)),
           let plain = try? AES.GCM.open(box, using: SymmetricKey(data: key)) {
            return plain
        }

        if iv.count == kCCBlockSizeAES128, let plain = aesCBCDecrypt(cipher, key: key, iv: iv) {
            return plain
        }
        throw ImportError.wrongPassword
    }

    static func pbkdf2SHA256(password: String, salt: Data, iterations: Int, length: Int) -> Data {
        var derived = Data(count: length)
        let passwordBytes = Array(password.utf8)
        _ = derived.withUnsafeMutableBytes { derivedPtr in
            salt.withUnsafeBytes { saltPtr in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    passwordBytes.map { Int8(bitPattern: $0) },
                    passwordBytes.count,
                    saltPtr.bindMemory(to: UInt8.self).baseAddress,
                    salt.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    UInt32(iterations),
                    derivedPtr.bindMemory(to: UInt8.self).baseAddress,
                    length
                )
            }
        }
        return derived
    }

    static func aesCBCDecrypt(_ cipher: Data, key: Data, iv: Data) -> Data? {
        var output = Data(count: cipher.count + kCCBlockSizeAES128)
        let outputCapacity = output.count
        var moved = 0
        let status = output.withUnsafeMutableBytes { outPtr in
            cipher.withUnsafeBytes { inPtr in
                key.withUnsafeBytes { keyPtr in
                    iv.withUnsafeBytes { ivPtr in
                        CCCrypt(
                            CCOperation(kCCDecrypt),
                            CCAlgorithm(kCCAlgorithmAES),
                            CCOptions(kCCOptionPKCS7Padding),
                            keyPtr.baseAddress, key.count,
                            ivPtr.baseAddress,
                            inPtr.baseAddress, cipher.count,
                            outPtr.baseAddress, outputCapacity,
                            &moved
                        )
                    }
                }
            }
        }
        guard status == kCCSuccess else {
            return nil
        }
        return output.prefix(moved)
    }

    // MARK: Helpers

    private static func string(_ dict: [String: Any], _ key: String) -> String {
        (dict[key] as? String) ?? ""
    }

    private static func int(_ dict: [String: Any], _ key: String) -> Int? {
        if let number = dict[key] as? Int {
            return number
        }
        if let text = dict[key] as? String {
            return Int(text)
        }
        return nil
    }

    static func firstNonEmpty(_ values: String?...) -> String {
        for value in values {
            if let trimmed = value?.trimmingCharacters(in: .whitespaces), !trimmed.isEmpty {
                return trimmed
            }
        }
        return ""
    }

    static func trimForDisplay(_ text: String) -> String {
        text.count <= 48 ? text : String(text.prefix(45)) + "..."
    }
}
